File: baekjoon/bfs/12851/Main.java
import java.util.*;

class Main {
    
    static ArrayList<ArrayList<Integer> > adj;
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int k = sc.nextInt();
        
        int l = n>k?n+2:k+2;
        adj = new ArrayList<ArrayList<Integer> >();
        for(int i=0; i<l; i++) {
            adj.add(new ArrayList<Integer>());
        }
        
        for(int i=0; i<l; i++) {
            if(i+1<l) adj.get(i).add(i+1);
            if(i-1>=0) adj.get(i).add(i-1);
            if(i*2<l) adj.get(i).add(i*2);
        }
        
        Queue<Integer> q = new LinkedList<Integer>();
        q.add(n);
        int [][] dist = new int[l][2];
        for(int i=0; i<l; i++) {
            dist[i][0] = -1;
            dist[i][1] = 0;
        }
        dist[n][0] = 0;
        dist[n][1] = 1;
        while(!q.isEmpty()) {
            int here = q.peek();
            q.remove();
            
            for(int i=0; i<adj.get(here).size(); i++) {
                int there = adj.get(here).get(i);
                if(dist[there][0] == -1) {
                    dist[there][0] = dist[here][0] + 1;
                    dist[there][1]++;
                    q.add(there);
                } else if(dist[there][0] == dist[here][0]+1) {
                    dist[there][1]++;
                    q.add(there);
                }
            }
        }
        
        System.out.println(dist[k][0] + "\n" + dist[k][1]);
    }
}

File: baekjoon/bfs/12851/main.cpp
#include <iostream>
#include <cstdio>
#include <vector>
#include <queue>

using namespace std;

vector<vector<int> > adj;
int main()
{
    int n,k;
    scanf("%d %d", &n, &k);
    adj = vector<vector<int> >((n>k?n+2:k+2));
    
    for(int i=0; i<adj.size(); i++) {
        if(i-1 >= 0) adj[i].push_back(i-1);
        if(i+1 < adj.size()) adj[i].push_back(i+1);
        if(i*2 < adj.size()) adj[i].push_back(i*2);
    }
    
    queue<int> q;
    q.push(n);
    vector<pair<int, int> > dist(adj.size(), make_pair(-1,0));
    dist[n].first = 0;
    dist[n].second = 1;
    while(!q.empty()) {
        int here = q.front();
        q.pop();
        
        for(int i=0; i<adj[here].size(); i++) {
            int there = adj[here][i];
            if(dist[there].first == -1) {
                dist[there].first = dist[here].first+1;
                dist[there].second++;
                q.push(there);
            } else if(dist[there].first == dist[here].first+1) {
                dist[there].second++;
                q.push(there);
            }
        }
    }
    
    printf("%d\n%d\n", dist[k].first, dist[k].second);
    return 0;
}